src/Models/uml_diagram.py

```diff
@@ -0,0 +1,109 @@
+from Models.uml_class import UML_Class
+from Models.uml_relation import UML_Relation, rel_types
+import re
+
+
+
+class UML_Diagram: 
+    
+    def __init__(self):
+        self._classes:list[UML_Class] = []
+        self._relations:list[UML_Relation] = []
+
+#===================================== Accessors =====================================#
+    def get_class(self, c_name:str) -> UML_Relation:
+        item = next((c for c in self._classes if c.get_name() == c_name), None)
+        if item is None: 
+            raise ValueError("Class %s does not exist" % c_name)
+        return item
+    
+    def get_relation(self, r_src:str, r_dst:str) -> UML_Relation:
+        #A relation between r_src and r_dst is the same as one between r_dst and r_src
+        item = next((r for r in self._relations if 
+                     r.get_src_name() == r_src and r.get_dst_name() == r_dst
+                     or r.get_src_name() == r_dst and r.get_dst_name() == r_src), None)
+        if item is None: 
+            raise ValueError("Relation between {0} and {1} does not exist.".format(r_src, r_dst))
+        return item
+        
+    def get_all_classes(self) -> list[UML_Class]:
+        return self._classes
+    
+    def get_all_relations(self) -> list[UML_Relation]:
+        return self._relations
+
+#===================================== Mutators =====================================#
+    def add_class(self, c_name:str) -> None:
```

Comment:
Seems more clear to do a contains check on the class list rather than try catching in attempt to get an error?

Reply:
Contains for objects appears to check memory locations. We cannot easily check for the existence of a string among a group of objects (technically we can do it using a generator that compares the names of the objects), but the logic would look something like ``` if (next ((x for x in list if x.get_name() == c_name), False): Raise ValueError ``` which isn't meaningfully cleaner than negating an existing method. 

One approach I had considered that would clean this up is adding an option to return None instead of raising an error (and defaulting to raising an error) which would result in simpler checks for these two methods but otherwise be unnecessary and risk not raising an error where one should be raised elsewhere in the code. 

This has been a long way of saying that I considered this option and tossed it because it necessitates duplicating the logic of get_class instead of just calling get_class.

Reply:
Here is a suggestion for revised more clear code although you don't have to implement it if you're dead set on try/catch

def add_class(self, c_name: str) -> None:
    if self.get_class(c_name) is not None:
        raise ValueError("Class %s already exists" % c_name)
    else:
        self._classes.append(UML_Class(c_name))